src/utilities/mta/range_proofs.rs

```diff
@@ -0,0 +1,718 @@
+        let z = (BigInt::mod_pow(h1, &a, N_tilde) * BigInt::mod_pow(h2, &ro, N_tilde)) % N_tilde;
+        let u = ((alpha.borrow() * &alice_ek.n + 1)
+            * BigInt::mod_pow(&beta, &alice_ek.n, &alice_ek.nn))
+            % &alice_ek.nn;
+        let w =
+            (BigInt::mod_pow(h1, &alpha, N_tilde) * BigInt::mod_pow(h2, &gamma, N_tilde)) % N_tilde;
+        Self {
+            alpha,
+            beta,
+            gamma,
+            ro,
+            z,
+            u,
+            w,
+        }
+    }
+}
+
+/// Represents the second round of the interactive version of the proof
+struct AliceZkpRound2 {
+    s: BigInt,
+    s1: BigInt,
+    s2: BigInt,
+}
+
+impl AliceZkpRound2 {
+    fn from(
+        alice_ek: &EncryptionKey,
+        round1: &AliceZkpRound1,
+        e: &BigInt,
+        a: &BigInt,
+        r: &BigInt,
+    ) -> Self {
+        Self {
+            s: (BigInt::mod_pow(r, &e, &alice_ek.n) * round1.beta.borrow()) % &alice_ek.n,
+            s1: (e * a) + round1.alpha.borrow(),
+            s2: (e * round1.ro.borrow()) + round1.gamma.borrow(),
+        }
+    }
+}
+
+/// Alice's proof
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct AliceProof {
+    z: BigInt,
+    e: BigInt,
+    s: BigInt,
+    s1: BigInt,
+    s2: BigInt,
+}
+
+impl AliceProof {
+    /// verify Alice's proof using the proof and public keys
+    pub fn verify(
+        &self,
+        cipher: &BigInt,
+        alice_ek: &EncryptionKey,
+        dlog_statement: &DLogStatement,
+    ) -> bool {
+        let N = &alice_ek.n;
+        let NN = &alice_ek.nn;
+        let N_tilde = &dlog_statement.N;
+        let h1 = &dlog_statement.g;
+        let h2 = &dlog_statement.ni;
+        let Gen = alice_ek.n.borrow() + 1;
+
+        if self.s1 > FE::q().pow(3) {
+            return false;
+        }
+
+        let z_e_inv = BigInt::mod_inv(&BigInt::mod_pow(&self.z, &self.e, N_tilde), N_tilde);
+        if z_e_inv.is_none() {
+            // z must be invertible, yet the check is done here
+            return false;
+        }
+        let z_e_inv = z_e_inv.unwrap();
+
+        let w = (BigInt::mod_pow(h1, &self.s1, N_tilde)
+            * BigInt::mod_pow(h2, &self.s2, N_tilde)
+            * z_e_inv)
+            % N_tilde;
+
+        let gs1 = (self.s1.borrow() * N + 1) % NN;
+        let cipher_e_inv = BigInt::mod_inv(&BigInt::mod_pow(&cipher, &self.e, NN), NN);
+        if cipher_e_inv.is_none() {
+            return false;
+        }
+        let cipher_e_inv = cipher_e_inv.unwrap();
+
+        let u = (gs1 * BigInt::mod_pow(&self.s, N, NN) * cipher_e_inv) % NN;
+
+        let e = HSha256::create_hash(&[N, &Gen, cipher, &self.z, &u, &w]);
+        if e != self.e {
+            return false;
+        }
+
+        true
+    }
+    /// Create the proof using Alice's Paillier private keys and public ZKP setup.
+    /// Requires randomness used for encrypting Alice's secret a.
+    /// It is assumed that secp256k1 curve is used.
```

Comment:
where do we assume this?

Reply:
We use FE::q() to create Alice's round 1. At first, I wanted to make MtA independent of the curve, passing its order as a parameter. But then we'd have to change hashing function too, and I thought it would be a bad idea to make users think that any group order can be passed while still using SHA256 inside. Thus the explicit note about `secp256k1`.
If you think it's important to make MtA independent of the curve, it's probably easy, the curve just would need to implement a trait that outputs a hash from Z_q where q is (at least approximately) the order of the curve.